### ISSUE TITLE
Inclusive Duration Interval

### ISSUE THREAD

**Issue by devonbuckingham** [nodatime/nodatime#1547]
One issue I have run into using NodaTime is that the duration of an interval is from the number of ticks in the interval, rather than the number of ticks in the interval _plus one_.

For example, if I had an interval whose start was the first tick of May 22nd, 2020, and whose end was the last tick of May 22nd, 2020, then the duration of that interval would _not_ be equal to `Duration.FromDays(1)`; it would be equal to `Duration.FromDays(1) - Duration.Epsilon`. In this case, I would likely want the duration from the _beginning_ of the first tick to the _end_ of the last tick, rather than the number of ticks.

Is there a design reason for this? Is there motivation to add something like `Interval.DurationInclusive`?

**Comment by malcolmr** [comment]
I'm possibly confused by what you're asking, but a (bounded) `Interval` (in Noda Time) is defined as a left-closed, right-open interval, so the `End` instant is never part of the interval (i.e. `interval.Contains(interval.End)` is always false).

So if you have an interval which covered all the instants for today (in UTC), it'd be expressed as `[2020-05-22T00:00:00Z, 2020-05-23T00:00:00Z)`, and would have a duration (correctly) of `Duration.FromDays(1)`.

If you set the end tick to the last tick of 2020-05-22 instead, then than tick won't be included in the interval, and so the duration of the interval would indeed be `Duration.FromDays(1) - Duration.Epsilon`.

**Comment by jskeet** [comment]
What Malcolm said, basically. Another way of thinking about it is to consider an *empty* interval. Consider this:

```csharp
Instant instant = ...; // It doesn't matter
Interval empty = new Interval(instant, instant);
```

That starts and ends at the same instant, and has a duration of 0. If we made it *include* `instant`, then it should have a duration of 1ns, because it would include that instant and no other.

Another way of thinking about it is to consider arrays. If you create an array like this:

```csharp
int[] values = new int[5];
```

... then it has a length of 5, and contains elements 0, 1, 2, 3 and 4. It *doesn't* have an element 5. In the same way, if you create an interval of `x` to `x.PlusNanoseconds(5)` it includes `xns`, `x+1ns`, `x+2ns`, `x+3ns` and `x+4ns`, but not `x+5ns`.

Basically it sounds like you should be constructing your intervals bearing in mind that the end is exclusive - so don't build an interval with an end point of "the last tick of 2020-05-22". In fact, this provides an example of another good reason for being half-open. What is "the last tick of 2020-05-22? Is it 2020-05-22T23:59:59.999999900? If so, the last 99 nanoseconds would be missing even if we included the nanosecond you specified. By expression the end point in terms of "the instant at which the interval ends" rather than "the final instant in the interval", the precision used is unimportant.

**Comment by devonbuckingham** [comment]
Makes sense. I just overlooked this when reading the documentation. Thanks!

**Comment by devonbuckingham** [comment]
Hey, I actually have a related question about the empty interval case that you described. I would have expected an empty interval to have a zero duration, which it does, but I also would have expected it to include its starting/ending `Instant`, which it does not. I think the key difference here is that I imagine an `Instant` as simply a zero-dimension point in time with no duration, whereas your description above implies that an instant has a duration of 1ns. However, everything I see in the NodaTime documentation describes an instant solely as representing a point in time, and nothing about it being considered to have a duration. Given this, wouldn’t it be more consistent and semantically sensible for the inclusivity of the starting instant of `Interval` to supersede the exclusivity of the ending instant, rather than the other way around? It seems odd to consider an `Instant` as having an intrinsic duration.

**Comment by jskeet** [comment]
> your description above implies that an instant has a duration of 1ns

No, an instant doesn't have a duration. But if an interval contains an instant, it logically means it contains every point in time from that instant up (but not including) the next representable instant.

> Given this, wouldn’t it be more consistent and semantically sensible for the inclusivity of the starting instant of Interval to supersede the exclusivity of the ending instant, rather than the other way around?

No, because that's not how empty intervals work in maths. An empty interval is precisely that: *empty*. Meaning it doesn't contain anything.

The current behavior of empty intervals achieves all kinds of mathematical consistencies that would be violated if `[a, a)` were deemed to include `a`. For example, the duration of an interval in nanoseconds is exactly matched by "the number of distinct `Instant` values where `interval.Contains(instant)`. With your proposal, that would be violated by the empty interval (which would have a duration of 0ns, but contain 1 instant). Thinking of an interval as a set, you're effectively saying that `[a, a)` should have the same containment predicate as `[a, a+1)` - namely to include `a` and only `a`.

Closing again, as the design is as intended and follows normal maths.
